### DISCUSSION TITLE
Berry isuse with NOT receiving MQTT when used as autoexec.be

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#15853] [Q&A] [answered]
My program below works fine, when I installed it via the Berry Scripting Console, I can send an MQTT message to it, work fine..
But if I upload it to the file system, change its name to autoexec.be, reboot, most everthing works except MQTT messages are NOT received??

If I reboot the Tasmota device, I see a request for a connection to my MQTT server but donot see the Berry request...

any clue where I should look...

This is the subscribe message when in console mode... not received in autoexec.be mode
~~~
Jun 23 12:11:36 centos7 mosquitto: 1656011496: Received SUBSCRIBE from DVES_6C4EFC
Jun 23 12:11:36 centos7 mosquitto: 1656011496: #011RSF/JUKEBOX/# (QoS 0)
~~~

~~~


#- 
   To load this file, compile Tasmota32 with the option as needed....
   Then Load the new binary image in your ESP32 and re-boot it. 
   Open the web page for this device, select Console, then Manage File System
   Rename this Berry file to "autoexec.be", then upload it to the ESP32 file system. 
   Reboot Tasmota, this Berry file will run after re-booting.
   
 -#
   
 #- *************************************** -#
 #-
    CHANGE LOG:
 
    DATE         REV  DESCRIPTION
    -----------  ---  ----------------------
    21-Jun-2022  1.0  TRL - 1st release
       
    Notes:  1)  Tested with 11.1.0.3(tasmota)
        
    ToDo:   1)

    tom@lafleur.us


    This is a jukebox player that get tracks to play from the Seeburg 3WA wallbox via driver (xsns_123_SB-3Wx)
    or via MQTT. It requires a change to the MP3 driver (xdrv_14_mp3) to sense an external busy siginal from 
    the MP3 module. This is then made avilable to Berry as a rule to delay next track selection untll MP3 module is not busy
    
    If you select V10 (index = 200), it toggles random play, on, off, if you select anything else, it reset 
    random play and start to play that selection. V9 (index = 199), will stop play and reset system.

    Will respond to MQTT data for Track, Volume etc... -->  subscribe('RSF/JUKEBOX/#')
 
-#

import string
import math
import mqtt
import json

#- *************************************** -#
#- *************************************** -#
class SEEBURG_DRIVER : Driver
    
    #build an global array-->list to store wallbox selection
    static buf = []
    var BusyFlag           # 0 = busy
    var RandomPlay

#- *************************************** -#   
    def init()
        
        tasmota.add_rule ("Wallbox",   /MyObj ->  self.process_wallbox(MyObj) )
        tasmota.add_rule ("MP3Player", /MyObj2 -> self.process_MP3_busy(MyObj2) )

        self.BusyFlag    = 0                               #  0 = busy
        self.RandomPlay  = false

        math.srand( int (tasmota.rtc()['utc'] ))           # seed random number

        mqtt.subscribe('RSF/JUKEBOX/#')  
        
        tasmota.cmd("MP3Volume 80")                        # set volume
        tasmota.cmd("MP3EQ 3")                             # set EQ
        self.buf.clear()                                   # flush the queue

    end

    #- *************************************** -#
    def queue(index)

        var MaxQueueSize = 10

        if (index > 200) index = 200 end                    # do a bounds check
        if (index <  0)  index = 0   end

        if (index == 200)                                   # if we have a selection of: V10 index = 200, toggle random play                  
            if (self.RandomPlay == true )  self.RandomPlay = false end
            if (self.RandomPlay == false ) self.RandomPlay = true  end
            self.buf.clear()                                # flush the queue            
            return                                          # exit function
        end

        if (index == 199)                                   # if we have a track of: V9 index = 199. reset
            tasmota.cmd("MP3Reset")
            self.buf.clear()                                # flush the queue 
            self.RandomPlay  = false
            tasmota.cmd("MP3Volume 80")                     # set volume
            return
        end

        self.RandomPlay = false                             # reset random play if we select anything else
        if (self.buf.size() >= MaxQueueSize) return  end    # do not add track selection to queue if full
        #print("Selection Index: ", index)

        self.buf.push(index)                                # add selection to list
        print ("Queue: ", self.buf)                         # print queue
    end

#- *************************************** -# 
    def mqtt_data(topic, idx, strdata, bindata)

    #print("Topic: ", topic) 
    var MyCmd

        if ( string.find(topic, "Track") > 0) 
            var MQTT_Index = json.load(strdata)
            if (MQTT_Index > 200) MQTT_Index = 200 end
            if (MQTT_Index <  0)  MQTT_Index = 0   end
            self.queue(MQTT_Index)                          # send to queue
            return true
        end
      
        if ( string.find(topic, "Volume") > 0) 
            var MyVolume = json.load(strdata)
            if (MyVolume > 100) MyVolume = 100 end
            if (MyVolume <  0)  MyVolume = 0   end
            MyCmd = string.format("MP3Volume %u", int (MyVolume))
            tasmota.cmd(MyCmd)                              # set volume
            return true
        end

        if ( string.find(topic, "EQ") > 0) 
            var MyEQ = json.load(strdata)
            if (MyEQ > 5)  MyEQ  = 5   end
            if (MyEQ < 0)  MyEQ  = 0   end
            MyCmd = string.format("MP3EQ %u", int (MyEQ))
            tasmota.cmd(MyCmd)                              # set command
            return true
        end

        # Folder format uses folder to store track, starting at folder 01 -> 99,
        # each folder can have 1->254 songs, folder 00 is not alowed, nor is track 0
        # so, mininmal value is 256 + 1 = 257, max is  99 * 256 = 25,344 + 254 = 25,598
        # used caution, experimental at this time!
        if ( string.find(topic, "Folder") > 0) 
            var MyFolder = json.load(strdata)
            if (MyFolder > 25598)  MyFolder  = 257 end      # we use 257 at folder 1, track 1
            if (MyFolder < 257)    MyFolder  = 257 end
            MyCmd = string.format("MP3Folder %u", int (MyFolder))
            tasmota.cmd(MyCmd)                              # set command
            return true
        end

        if ( string.find(topic, "Reset") > 0) 
            tasmota.cmd("MP3Reset")
            self.buf.clear()                                # flush the queue 
            tasmota.cmd("MP3Volume 80")                     # set volume
            tasmota.cmd("MP3EQ 3")                          # set EQ
            return true 
        end

        if ( string.find(topic, "Pause") > 0) tasmota.cmd("MP3Pause") return true end
        if ( string.find(topic, "Play")  > 0) tasmota.cmd("MP3Play")  return true end
        if ( string.find(topic, "Stop")  > 0) tasmota.cmd("MP3Stop")  return true end

        return false
    end


#- *************************************** -# 
    def process_MP3_busy(MyObj2) 

        if  MyObj2 == nil print("Bad Obj2")     return end
        if !(MyObj2.contains('MP3Busy'))        return end
        self.BusyFlag = real(MyObj2['MP3Busy'] )                # get busy flag from MP3 driver
    end

#- *************************************** -# 
    def process_wallbox(MyObj)

        if  MyObj == nil print("Bad Obj")       return end
       
        #if !(MyObj.contains('Number_Index'))    return end
        #if !(MyObj.contains('Letter_Index'))    return end
        if !(MyObj.contains('Selection_Index')) return end      
            
        #var num = real(MyObj['Number_Index'] )
        #var let = real(MyObj['Letter_Index'] )
        var index = real(MyObj['Selection_Index'] )             # get track selection from Wallbox
        
        self.queue(index)                                       # send to queue  
    end


#- *************************************** -#
    def play(Index)

        if (self.BusyFlag == 1)                                  # if not busy...
            print ("Playing Track: ", Index)
            var MyCmd = string.format("MP3Track %u", int (Index))
            tasmota.cmd(MyCmd)
            print ("Queue: ", self.buf)                          # print queue
        end
    end

#- *************************************** -#
    def every_second()

        if (self.RandomPlay == true )                           # select a random track, range 1 -> 198      
            if (self.BusyFlag == 1)                             # if not busy...
                var random = math.rand() % (198 + 1 - 1) + 1    # rand() % (Max + 1 - Min) + Min
                print ("Random Track: ", random)
                self.play(random)
                return
            end
        end
    
        if (self.buf.size() == 0) return  end                   # nothing to do, selection queue is empty
        if (self.BusyFlag == 1)                                 # if not busy...
            var NextTrack = self.buf.pop(0)                     # get next track from stack
            self.play(NextTrack)
        end
    end
  
   
#- *************************************** -#
    def web_sensor()
        #tasmota.web_send_decimal(msg)
    end
  

#- *************************************** -#
    def json_append()
        #tasmota.response_append(msg)     
    end
  

end     # end of class SEEBURG_Driver : Driver


#- *************************************** -#
SEEBURG_Driver =   SEEBURG_DRIVER()
tasmota.add_driver(SEEBURG_Driver)

#- ************ The Very End ************* -#


~~~

**Comment by sfromis** [comment]
MQTT subscribe can only work when Tasmota is connected to MQTT, which it not is at autoexec time. I really do not recommend the technique of renaming a program file to autoexec.be, I'd rather have that startup script have short calls to activate whatever I need, possibly delayed to an appropriate time.

Using `tasmota.add_rule("mqtt#connected", ...)` is what I usually do. Beware that the subscription status will be lost in case of the MQTT connection being restarted for some reason. Thus, using `tasmota.add_rule("mqtt#disconnected", ...)` may be relevant too.

One bug which hopefully soon will be fixed is that the handler for the subscribe is not removed for a `mqtt.unsubscribe()`, meaning that if you subscribe to the same topic again, both the new and old handler are called.

**Comment by trlafleur** [comment]
do you have an example of what you recommend by a short script to invoke
another Berry program??

thanks

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
Simplest would just be a `load('myprogram.be')`, if the loading is not to be delayed.

The already suggested `tasmota.add_rule("mqtt#connected", ...)` could be one way of having a relevant delay.

First, you do need to have a strategy for how you want to handle subscriptions, as you should be prepared to resubscribe later, as already pointed out.

**Reply by trlafleur** [reply]
I made the below change to my code, MQTT connects returns a 1, again, works fine in console mode...

~~~
1 | MQTT_CONNECT_BAD_PROTOCOL | the server doesn't support the requested version of MQTT
~~~

~~~
#- *************************************** -#   
    def init()
    
        import mqtt
 
        tasmota.add_rule ("Wallbox",            /MyObj ->  self.process_wallbox(MyObj) )
        tasmota.add_rule ("MP3Player",          /MyObj2 -> self.process_MP3_busy(MyObj2) )
        tasmota.add_rule ("mqtt#connected",     /MyObj3 -> self.connect_MQTT(MyObj3)) 
        tasmota.add_rule ("mqtt#disconnected",  /MyObj4 -> self.disconnected_MQTT(MyObj4))

        self.BusyFlag    = 0                               #  0 = busy
        self.RandomPlay  = false

        math.srand( int (tasmota.rtc()['utc'] ))           # seed random number

        mqtt.subscribe('RSF/JUKEBOX/#')  
        
        tasmota.cmd("MP3Volume 80")                        # set volume
        tasmota.cmd("MP3EQ 3")                             # set EQ
        self.buf.clear()                                   # flush the queue

    end
    
 
 #- *************************************** -#    
    def disconnected_MQTT(MyObj4)
    
        print ("MQTT Disconnect: ",MyObj4)
    
    end
    
    
    
#- *************************************** -#   
    def connect_MQTT(MyObj3)
    
        print("Mqtt Connect: ",MyObj3)
        
        #mqtt.subscribe('RSF/JUKEBOX/#')  
        
        #tasmota.cmd("MP3Volume 80")                        # set volume
        #tasmota.cmd("MP3EQ 3")                             # set EQ
        #self.buf.clear()                                   # flush the queue
    
    end
    
    ~~~

**Comment by trlafleur** [comment]
also, if I do get a disconnected message, do I need to just resubscribe and
wait for a connected message??

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
As already pointed out, subscribe cannot work unless a MQTT connection exits for the broker to receive the subscription. Tasmota is not trying to queue anything awaiting a connection being established.

**Comment by s-hadinger** [comment]
Oh man. Handling of Mqtt subscriptions is too barebone. I need to fix unsubscribe and see if I can provide a simpler way, hiding all the required rules

**Comment by trlafleur** [comment]
Yes, I believe I understand the issues?!

For now, I’m just trying to understand what is return from the connect
callback.     Is a 1 telling me it connect? So that I can subscribe?  Or
something else??


On Thu, Jun 23, 2022 at 2:16 PM sfromis ***@***.***> wrote:

> As already pointed out, subscribe cannot work unless a MQTT connection
> exits for the broker to receive the subscription. Tasmota is not trying to
> queue anything awaiting a connection being established.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/15853#discussioncomment-3013559>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC4EK6YDM3SC6Z5DNM5MP3VQTH2HANCNFSM5ZVHWQVQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
It is not a connect callback. It is a connect**ed** callback, meaning that it **is** connected.

The value 1 is merely a constant. Ignore it.

**Comment by trlafleur** [comment, marked as answer]
Yes, spelling on my I phone is painful at times.


So when it returns a 1, it connect?



On Thu, Jun 23, 2022 at 2:50 PM sfromis ***@***.***> wrote:

> It is not a connect callback. It is a connect*ed* callback, meaning that
> it *is* connected.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/15853#discussioncomment-3013699>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC4EK6O3TKCNQ4HKTP6BMDVQTLZLANCNFSM5ZVHWQVQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
As I added, the value 1 is merely a constant. Ignore it, it has no meaning.

**Reply by sfromis** [reply]
Again, while your spelling was ok, it was me pointing out that the callback is not about doing a connect, but about reporting that it **already** has connected.

**Reply by trlafleur** [reply]
ok, did an autoexec.be that loads my program...  It waits for mqtt#connected, then load my program... look like this solved the issues...

None of this is very obvious at first glance...

~~~
#- *************************************** -#
#- *************************************** -#
class SEEBURG_LOADER : Driver

#- *************************************** -#   
    def init()
        import mqtt
        tasmota.add_rule ("mqtt#connected", /MyObj3 -> self.connected_MQTT(MyObj3)) 
    end

#- *************************************** -#   
    def connected_MQTT(MyObj3)
        #print("Mqtt Connected: ",MyObj3)
        load ('Seeburg1.be') 
        end
    end   
  
#- *************************************** -#
SEEBURG_Loader =   SEEBURG_LOADER()
tasmota.add_driver(SEEBURG_Loader)

#- ************ The Very End ************* -#
~~~

**Reply by sfromis** [reply]
You  could shorten all that to just one line:
`tasmota.add_rule('mqtt#connected', /->load('Seeburg1.be'))`

You need no class as you need no instance. You need no add_driver, as the loader has no driver functionality. You need no `import mqtt` as you do not use the mqtt module there. And the closure notation can be very convenient for doing small things.

Much easier to read too.

Beware that both your variant, and the shortened one, will load `Seeburg1.be` again if/when Tasmota reconnects to MQTT for whatever reason, which is pretty much within the realm of the expect-able.

Depending on other things, sometimes the following strategy can make sense, to mitigate the double-load issue:
```
tasmota.add_rule('mqtt#connected', 
  def ()
    load('Seeburg1.be')
    tasmota.remove_rule('mqtt#connected')
  end)
```

**Reply by barbudor** [reply]
You may be interested in testing SetOption140 1 which enables persistent sessions. So in case of disconnection and reconnection the broker keeps the context including the subscribed topics and queues any message that happened during the disconnection time

**Reply by sfromis** [reply]
FTR, the issue I mentioned with `mqtt.unsubscribe` not removing the handler, which would still be called if resubscribing to same topic, has now been fixed by s-hadinger, if you use the very latest dev version. And a no-argument unsubscribe will ditch all subscriptions from Berry. #15876

**Comment by trlafleur** [comment]
Will try that, I just cut and paste code to see if would work


Thanks


On Thu, Jun 23, 2022 at 4:29 PM sfromis ***@***.***> wrote:

> You could shorten all that to just one line:
> tasmota.add_rule('mqtt#connected', /->load('Seeburg1.be')
>
> You need no class as you need no instance. You need no add_driver, as the
> loader has no driver functionality. You need no import mqtt as you do not
> use the mqtt module there. And the closure notation can be very convenient
> for doing small things.
>
> Much easier to read too.
>
> Beware that both your variant, and the shortened one, will load
> Seeburg1.be again if/when Tasmota reconnects to MQTT for whatever reason,
> which is pretty much within the realm of the expect-able.
>
> Depending on other things, sometimes the following strategy can make
> sense, to mitigate the double-load issue:
>
> tasmota.add_rule('mqtt#connected',
>   def ()
>     load('Seeburg1.be')
>     tasmota.remove_rule('mqtt#connected')
>   end)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/15853#discussioncomment-3014013>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABC4EK25OGJ4TZH3YABUMV3VQTXPDANCNFSM5ZVHWQVQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 

~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by sfromis** [reply]
Note that with recent enhancements, you can use `mqtt.subscribe` in `autoexec.be` before MQTT is even connected, as `mqtt.subscribe` now is doing the subscription management for you, including resubscribing when relevant, like in case of network outages.
